apps/emqx/src/emqx_cm.erl

```diff
@@ -262,96 +262,64 @@ set_chan_stats(ClientId, ChanPid, Stats) ->
     {ok, #{
         session := emqx_session:session(),
         present := boolean(),
-        pendings => list()
+        replay => _ReplayContext
     }}
     | {error, Reason :: term()}.
 open_session(true, ClientInfo = #{clientid := ClientId}, ConnInfo) ->
     Self = self(),
-    CleanStart = fun(_) ->
+    emqx_cm_locker:trans(ClientId, fun(_) ->
         ok = discard_session(ClientId),
-        ok = emqx_session:destroy(ClientId),
+        ok = emqx_session:destroy(ClientInfo, ConnInfo),
         create_register_session(ClientInfo, ConnInfo, Self)
-    end,
-    emqx_cm_locker:trans(ClientId, CleanStart);
+    end);
 open_session(false, ClientInfo = #{clientid := ClientId}, ConnInfo) ->
     Self = self(),
-    ResumeStart = fun(_) ->
-        case takeover_session(ClientId) of
-            {living, ConnMod, ChanPid, Session} ->
-                ok = emqx_session:resume(ClientInfo, Session),
-                case wrap_rpc(emqx_cm_proto_v2:takeover_finish(ConnMod, ChanPid)) of
-                    {ok, Pendings} ->
-                        clean_register_session(Session, Pendings, ClientInfo, ConnInfo, Self);
-                    {error, _} ->
-                        create_register_session(ClientInfo, ConnInfo, Self)
-                end;
-            none ->
+    emqx_cm_locker:trans(ClientId, fun(_) ->
+        case emqx_session:open(ClientInfo, ConnInfo) of
+            {true, Session, ReplayContext} ->
+                ok = register_channel(ClientId, Self, ConnInfo),
+                {ok, #{session => Session, present => true, replay => ReplayContext}};
+            false ->
                 create_register_session(ClientInfo, ConnInfo, Self)
         end
-    end,
-    emqx_cm_locker:trans(ClientId, ResumeStart).
```

Comment:
Discussion point. @thalesmg @ieQu1 

Half of this stuff is now part of `takeover_channel_session/2` function, which [is used by `emqx_session_mem` itself](https://github.com/emqx/emqx/blob/b8184966ab40e847c2b0b5e2bf6399aa48fff0fc/apps/emqx/src/emqx_session_mem.erl#L207-L220) (since it's essentially part of in-memory session logic). That's why there's now a functor for "stuff that should happen after takover begin but before takeover end".

What was named `Pendings` before is now hidden behind opaque `ReplayContext` term, named that way because it [gets passed back to `emqx_session:replay/3` later](https://github.com/emqx/emqx/blob/b8184966ab40e847c2b0b5e2bf6399aa48fff0fc/apps/emqx/src/emqx_channel.erl#L2028).

This needs a closer look to ensure that logic is essentially the same as before. See also next comment.

Reply:
I still have to look more closely; as an initial observation we seem to have a chain of calls `emqx_cm` -> `emqx_session_mem` -> `emqx_cm` -> `emqx_session_mem` (via the continuation), and that only happens in the memory implementation.  Since that function is now only used by `emqx_session_mem`, perhaps the function could be entirely moved to `emqx_session_mem`.

Reply:
Good point, I think it indeed should ideally be there. I believe I thought about that but was stopped by the fact that the logic of `takeover_channel_session/2` heavily relies on `emqx_cm_proto_v2`, hence must stay in `emqx_cm`. However, we may try to end up in a middle ground: export `takeover_begin` / `takeover_end` from `emqx_cm`, and use them in `emqx_cm`. What do you think?

Reply:
I thought that the reason was to avoid calling `emqx_cm_provo_v2` outside `emqx_cm`.  But, IMO, making the flow harder to follow outweighs calling `emqx_cm_proto_v2` in `emqx_session_mem`.  

If the naming of the proto module feels weird, we could introduce `emqx_session_mem_proto_v1` that calls the same old `emqx_cm:takeover_{session,finish}` functions, as they will exist on older versions.

> However, we may try to end up in a middle ground: export takeover_begin / takeover_end from emqx_cm, and use them in emqx_cm. What do you think?

IIUC, `takeover_session` (begin) and `takeover_finish` are already exported since they are RPC targets, so they could be called from `emqx_session_mem`?